### DISCUSSION TITLE
ACL application micro-lc

### DISCUSSION THREAD

**Discussion by giovanni-motterle-mia-care** [mia-platform/community#539] [Q&A] [answered]
### What product are you having troubles with?

Microfrontend Composer

### What Console version are you using?

v13.0.5

### Description

Hi, I was reading about ACL [expessions in micro-lc ](https://micro-lc.io/add-ons/backend/middleware#acl-application), specifically the ones based on permissions.
Permissions ACLs are based on a request header that contains the permissions list. Isn't it a limitation using headers in this use case? I know headers size is limited (I think it's 8kb default on node and customizable).
Since I would prefer not using headers, i thought that [custom caller property extraction](https://micro-lc.io/add-ons/backend/middleware#custom-extraction) would come handy.
I don't understand how to create a "user-defined custom function" though. Moreover, is it possible to make HTTP calls inside of it?

Thanks

**Comment by epessina** [comment]
Hi @giovanni-motterle-mia-care , thanks for reaching out!

To provide a custom extractor you need to mount a JS module that exports your function in the directory specified by the `ACL_CONTEXT_BUILDER_PATH` [env variable](https://micro-lc.io/add-ons/backend/middleware/#environment-variables) (it defaults to `/usr/src/app/config/acl-context-builder.js`).

In Console, you need to create a new config map for your middleware instance with the JS file inside at the path specified in the variable above (if you use the default path remember to check the "Preserve files" box!).

The function exported by the file cannot make HTTP calls (its execution is sandboxed for safety reasons), so your permissions must be contained either in the headers or the query parameters. If you are hitting a size limit we can think of a solution

**Reply by giovanni-motterle-mia-care** [reply]
Got it, thanks.
I'm not hitting the size limit right now as I haven't implemented permission based ACLs yet but, header limits are so small that I think it is inevitable to reach the limit.
If from the module it would be possible to make http calls I could read the permissions directly from my DB without any problems and with great flexibility.

**Reply by epessina** [reply]
For safety reasons, I cannot see us adding the possibility of making HTTP calls from the module.
If the limit is hit my suggestion would be configuring the gateway accordingly.

**Reply by giovanni-motterle-mia-care** [reply]
So your suggestion is to extend headers size limit? I guess this should be configured on every microservice that use the permissions header and is not really a solution as size limit can be reached with new permissions creation.

**Reply by epessina** [reply]
The reason why we cannot add the possibility to make HTTP calls inside of the module is that the user defined code is executed in a sandbox, since it's untrusted. So it doesn't make sense to allow HTTP calls.

This leaves the headers as the most feasible route, in my opinion

**Reply by giovanni-motterle-mia-care** [reply]
I don't understand the necessity of not trusting code that an authorized developer is writing, why is that?
To me, the header mechanism is not appropriate for some serious production-grade applications.

**Reply by amountainram** [reply]
If you need to perform an HTTP request to enhance the context of the interpolation you are free to do it for instance with a side-car you could place near `middleware`. The request forwarded to `middleware` could be enhanced with known or custom headers or even modifying the request body, if you rather  do not want to use headers.

In terms of static serializable configuration of the interpolation, `middleware` already does the best it can and quite safely I'd add.
We recently featured the sandboxed to avoid creating a JSON-schema-like metalanguage to configure the request parsing and allow any manipulation on the HTTP request provided by a sandboxed global, namely without `eval` and `fetch`. The security concerns are self-explanatory. For instance avoid (EVEN BY MISTAKE) to send user tokens to a third-party domain.

IMO it's even too much permissive since you can log from the sandbox and could leak cookies and other stuff.

In terms of security a good rule of thumb is "think always at the worst case" where here could be you are allowed to fetch a JS file and run it in the backend without checks. Needless to say how dangerous can be when dealing, for instance, with sensitive data. Obviously you could even post the HTTP request plaintext body which in some cases is even worse than the user token.

Regarding your statement:
> the header mechanism is not appropriate for some serious production-grade applications

wouldn't you say that the `Authorization` Bearer token just ENRICHES an HTTP request with user information + permission within the signed JWT token? User context + Authz context is then provided via header.

The info you gave about node limits was correct up to version 14 when it was moved to 16kB. As reference a quite sizable request with authorization header an a long token requires 2/3 kB. Well within ranges. Anyway when you run `middleware` you could add `NODE_OPTIONS=--max-http-header-size=<VALUE IN kB>` as env var to further extend the limit.

**Reply by giovanni-motterle-mia-care** [reply]
If I modify the request body forwarded to middleware, then are you sure that I can use it in the custom ACL context builder? 
From the [doc](https://micro-lc.io/add-ons/backend/middleware#example-1) I see

`
export default ({ headers, method, pathParams, queryParams, url }) => {
  const property = headers['my-custom-header']
  return [property]
}
`

Can I also write `{ body, headers, method, pathParams, queryParams, url })`?

**Reply by epessina** [reply]
No, the request body is not a parameter of the function at the moment. The reason is twofold:
- all the routes exposed by middleware are GET (and GET requests are not supposed to have a body with a semantic meaning)
- if the body is a stream, it should be consumed completely in the function and then re-streamed defiling the usefulness of having a stream in the first place

Having said that for us it's not a problem to add the body to the function since the implementation is trivial, if you think it can be useful in your scenario

**Reply by giovanni-motterle-mia-care** [reply]
Yes, I think I'm going to use it. Are you going to expose also a POST endpoint then?

**Reply by epessina** [reply]
No, we will leave them as GET, since it wouldn't have any meaning for them to be POSTs.

Moreover, it is not forbidden for GETs to have a body, it's just formally not recommended. And since this is a particularly use case of yours and the body is actually used in your own module (and thus doesn't carry a semantic meaning for the request itself) I don't see a problem in providing a body for your GET requests.

We will track the activity and keep you posted on the development

**Reply by giovanni-motterle-mia-care** [reply]
Thank you!

**Reply by giovanni-motterle-mia-care** [reply]
Hi, do you have an idea of when this new feature will be supported?
Thanks

**Reply by giovanni-motterle-mia-care** [reply]
Hi, any updates on this? We are starting working with permissions in backoffice and for our use case we are already counting roles with a set of permissions with over 10k characters total

**Reply by epessina** [reply]
Hi @giovanni-motterle-mia-care, sorry for the late reply

Unfortunately, we won't be able to address this issue for another two or three weeks. If you need it sooner, feel free to open a PR to [middleware](https://github.com/micro-lc/middleware). In particular, you need to work on the [extact-acl.ts](https://github.com/micro-lc/middleware/blob/main/src/lib/extract-acl-context.ts#L31) file to add the `body` of the request to the ACL context.

**Reply by giovanni-motterle-mia-care** [reply]
Hi, we are actually implementing a work-around to reduce the number of permissions of a user. 
We are forwarding to middleware the intersections of permissions used by the particular configuration page the user is requesting and the user permissions.
In this way we are not reaching headers size limit.
Despite that, I think it's still a useful feature to add when you can.

**Reply by epessina** [reply, marked as answer]
Hi @giovanni-motterle-mia-care . For your information, version `3.3.1` of [middleware](https://github.com/micro-lc/middleware) is out with support to `request.body` as ACL context builder input.

**Reply by giovanni-motterle-mia-care** [reply]
Hi @epessina, thank you !
